### PR TITLE
Write a documentation about auto adapt bs and auto adapt num_workers

### PR DESCRIPTION
**Pull request by eunwoosh**
### Summary
Write a documentation about auto adapt bs and auto adapt num_workers

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: docs/source/guide/explanation/additional_features/auto_configuration.rst [hunk 1]
```diff
@@ -55,3 +55,59 @@ Move to this folder and simply run without any options to start training:
...
     $ otx train
+
+
+Auto-adapt batch size
+---------------------
+
+This feature adapts a batch size based on the current hardware environment.
+There are two methods available for adapting the batch size.
+
+1. Checking GPU Compatibility
```
Comment (goodsong81) [inline docs/source/guide/explanation/additional_features/auto_configuration.rst]:
```suggestion
1. Prevent GPU Out of Memory (`Safe` mode)
```

File: docs/source/guide/explanation/additional_features/auto_configuration.rst [hunk 2]
```diff
@@ -55,3 +55,59 @@ Move to this folder and simply run without any options to start training:
...
+To address this, a maximum batch size that is still practical is provided.
+The learning rate is also adjusted based on the updated batch size.
+
+To enable this feature, add the following command:
+
+.. code-block::
+
+    $ otx train params --learning_parameters.auto_adapt_batch_size Safe
+
+2. Finding a possible large Batch Size
```
Comment (goodsong81) [inline docs/source/guide/explanation/additional_features/auto_configuration.rst]:
```suggestion
2. Find the maximum executable batch size (`Full` mode)
```

File: docs/source/guide/explanation/additional_features/auto_configuration.rst [hunk 3]
```diff
@@ -55,3 +55,59 @@ Move to this folder and simply run without any options to start training:
...
+
+The second method aims to find a possible large batch size that reduces the overall training time.
+Increasing the batch size reduces the number of iterations required for validation, thus speeding up training.
```
Comment (goodsong81) [inline docs/source/guide/explanation/additional_features/auto_configuration.rst]:
```suggestion
Increasing the batch size reduces the effective number of iterations required to sweep the whole dataset, thus speeds up the end-to-end training.
```

File: docs/source/guide/explanation/additional_features/auto_configuration.rst [hunk 4]
```diff
@@ -55,3 +55,59 @@ Move to this folder and simply run without any options to start training:
...
+1. Checking GPU Compatibility
+
+The first method checks if the current batch size is compatible with the available GPU devices.
+Larger batch sizes consume more GPU memory for training. Therefore, the system verifies if training is possible with the current batch size.
+If it's not feasible, the batch size is decreased to reduce GPU memory usage.
+However, setting the batch size too low can slow down training.
+To address this, a maximum batch size that is still practical is provided.
+The learning rate is also adjusted based on the updated batch size.
+
+To enable this feature, add the following command:
```
Comment (goodsong81) [inline docs/source/guide/explanation/additional_features/auto_configuration.rst]:
```suggestion
To enable this feature, add the following parameter:
```

File: docs/source/guide/explanation/additional_features/auto_configuration.rst [hunk 5]
```diff
@@ -55,3 +55,59 @@ Move to this folder and simply run without any options to start training:
...
+    $ otx train params --learning_parameters.auto_adapt_batch_size Safe
+
+2. Finding a possible large Batch Size
+
+The second method aims to find a possible large batch size that reduces the overall training time.
+Increasing the batch size reduces the number of iterations required for validation, thus speeding up training.
+However, it does not search for the maximum batch size as it is not efficient and may require significantly more time without providing substantial acceleration compared to a large batch size.
+Similar to the previous method, the learning rate is adjusted according to the updated batch size.
+
+To use this feature, include the following command:
```
Comment (goodsong81) [inline docs/source/guide/explanation/additional_features/auto_configuration.rst]:
```suggestion
To use this feature, add the following parameter:
```

File: docs/source/guide/explanation/additional_features/auto_configuration.rst [hunk 6]
```diff
@@ -55,3 +55,59 @@ Move to this folder and simply run without any options to start training:
...
+Auto-adapt num_workers
+----------------------
+
+This feature adapts the ``num_workers`` parameter based on the current hardware environment.
+The ``num_workers`` parameter controls the number of subprocesses used for data loading during training.
+While increasing ``num_workers`` can reduce data loading time, setting it too high can consume a significant amount of CPU memory.
+
+To simplify the process of setting ``num_workers`` manually, this feature automatically determines the optimal value based on the current hardware status.
+
+To enable this feature, use the following command:
```
Comment (goodsong81) [inline docs/source/guide/explanation/additional_features/auto_configuration.rst]:
```suggestion
To enable this feature, add the following parameter:
```

File: docs/source/guide/explanation/additional_features/auto_configuration.rst [hunk 7]
```diff
@@ -55,3 +55,59 @@ Move to this folder and simply run without any options to start training:
...
+There are two methods available for adapting the batch size.
+
+1. Checking GPU Compatibility
+
+The first method checks if the current batch size is compatible with the available GPU devices.
+Larger batch sizes consume more GPU memory for training. Therefore, the system verifies if training is possible with the current batch size.
+If it's not feasible, the batch size is decreased to reduce GPU memory usage.
+However, setting the batch size too low can slow down training.
+To address this, a maximum batch size that is still practical is provided.
+The learning rate is also adjusted based on the updated batch size.
```
Comment (goodsong81) [inline docs/source/guide/explanation/additional_features/auto_configuration.rst]:
```suggestion
To address this, the batch size is reduced to the maximum amount that could be run safely on the current GPU resource.
The learning rate is also adjusted based on the updated batch size accordingly.
```

**Review by goodsong81 — DISMISSED** [review]
Thank you for the update. It looks good to me in general.
I have some suggestions, but please feel free to apply or not at your discretion.

**Comment by eunwoosh** [comment]
> Thank you for the update. It looks good to me in general. I have some suggestions, but please feel free to apply or not at your discretion.

Thanks for your comment! I applied all of your comments :)

**Review by goodsong81 — APPROVED** [review]

**Review by sungmanc — APPROVED** [review]

**Pull request merged**